src/main/java/org/broadinstitute/consent/http/service/UserService.java

```diff
@@ -454,12 +454,135 @@ public void hasValidActiveERACredentials(User user) {
+    boolean hasInstitutionMatchingEmailDomain =
+        hasInstitutionMatchingEmailDomain(institutionFromEmail);
+
+    boolean modifiedUser = false;
+
+    if (hasInstitutionMatchingEmailDomain) {
+      if (handleUserWithInstitutionInMap(user, institutionFromEmail, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    } else {
+      if (handleUserWithoutInstitutionInMap(user, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    }
+
+    if (modifiedUser) {
+      return findUserByEmail(user.getEmail());
+    } else {
+      return user;
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean handleUserWithInstitutionInMap(User user, Institution institutionFromEmail, Institution institutionFromDatabase) {
+    boolean modifiedUser = false;
+    if (!hasMatchingInstitutionInDatabase(institutionFromEmail, institutionFromDatabase)) {
+      assignInstitutionToUser(user, institutionFromEmail.getId());
+      // Set these so we don't have to look the user up again in the database for the next bit of
+      // work with the library card.
+      user.setInstitutionId(institutionFromEmail.getId());
+      user.setInstitution(institutionFromEmail);
+      modifiedUser = true;
+    }
+
+    if (handleLibraryCardForUser(user)) {
+      modifiedUser = true;
+    }
+
+    return modifiedUser;
+  }
+
+  @VisibleForTesting
+  protected boolean handleLibraryCardForUser(User user) {
+    boolean modifiedUser = false;
+    if (hasLibraryCard(user)) {
+      try {
+        User lcIssuer = findUserById(user.getLibraryCards().get(0).getCreateUserId());
+        Institution lcIssuerInstitution = institutionService.findInstitutionForEmail(lcIssuer.getEmail());
+        if (!user.getInstitution().equals(lcIssuerInstitution)) {
+          dropLibraryCardForUser(user);
+          modifiedUser = true;
+        }
+      } catch (NotFoundException nfe) {
+        dropLibraryCardForUser(user);
+        modifiedUser = true;
+      }
+    }
+    return modifiedUser;
+  }
+
+  @VisibleForTesting
+  protected boolean handleUserWithoutInstitutionInMap(User user, Institution institutionFromDatabase) {
+    if (hasLibraryCard(user)) {
+      dropLCAndInstitutionForUser(user);
+      return true;
+    } else {
+      if (institutionFromDatabase != null) {
+        assignInstitutionToUser(user, null);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void assignInstitutionToUser(User user, Integer institutionId) {
+    userDAO.updateInstitutionId(user.getUserId(), institutionId);
+  }
+
+  private void dropLCAndInstitutionForUser(User user) {
+    assignInstitutionToUser(user, null);
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+  }
+
+  private void dropLibraryCardForUser(User user) {
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+  }
+
+  @VisibleForTesting
+  protected boolean hasInstitutionMatchingEmailDomain(Institution institution) {
+    return institution != null;
+  }
```

Comment:
Naming here is confusing. I would assume this would mean that the method tells you if an institution and an email are a match or not.